### PR TITLE
PHP 8.2 Support: Readonly classes

### PR DESCRIPTION
**Pull request by junichi11**
#4725
- https://wiki.php.net/rfc/readonly_classes

#### Part 1

- Fix the grammar file
- Add `PHP82UnhandledError`
- Add/Fix unit tests for the parser, the lexter, and the index

#### Part 2

- Fix hints for readonly classes
- Add the code template for a readonly class
- Add unit tests for hints, formatter, and CC

nb-php82-readonly-classes-syntax-and-hints

nb-php82-readonly-classes-syntax-and-hints2

### CODE REVIEW

**Comment by junichi11** [comment]
@tmysik Could you please have a look at this when you have time? If there is no problem, let's merge it :) Thanks!

**Comment by junichi11** [comment]
@tmysik Will increment the spec version.

File: php/php.editor/src/org/netbeans/modules/php/editor/model/nodes/ClassDeclarationInfo.java [hunk 1]
```diff
@@ -93,14 +94,24 @@ public Set<QualifiedName> getInterfaceNames() {
...
+                    phpModifiers.add(PhpModifiers.ABSTRACT);
+                    break;
+                case FINAL:
+                    phpModifiers.add(PhpModifiers.FINAL);
+                    break;
+                case READONLY:
+                    phpModifiers.add(PhpModifiers.READONLY);
+                    break;
+                default:
+                    break;
```
Comment (tmysik) [inline php/php.editor/src/org/netbeans/modules/php/editor/model/nodes/ClassDeclarationInfo.java:111]: Don't we want to log/throw here?

**Review by tmysik — APPROVED** [review]

File: php/php.editor/src/org/netbeans/modules/php/editor/model/nodes/ClassDeclarationInfo.java (hunk 1 above)
Comment (junichi11) [reply php/php.editor/src/org/netbeans/modules/php/editor/model/nodes/ClassDeclarationInfo.java:111]: Will add the following. Thanks!
```java
                case NONE:
                    // no-op
                    break;
                default:
                    assert false : "Handle " + modifier + " modifier";
                    break;
```

**Comment by junichi11** [comment]
@tmysik Thank you for your review! I'll fix it with the next PR.

File: php/php.editor/src/org/netbeans/modules/php/editor/model/nodes/ClassDeclarationInfo.java (hunk 1 above)
Comment (tmysik) [reply php/php.editor/src/org/netbeans/modules/php/editor/model/nodes/ClassDeclarationInfo.java:111]: 👍

**Comment by tmysik** [comment]
@junichi11 Nothing to thank for, really. Plenty of tests really help us here. Thanks for your great work!

**Pull request merged**